cranelift-codegen/src/peephole.rs

```diff
@@ -0,0 +1,469 @@
+// binding already appears in the environment.  I implemented this, but it
+// seems almost entirely ineffective, so I removed it again.
+
+// This describes a spill slot in the obvious way.  It also tracks the SSA
+// value name associated with the slot.
+#[derive(Clone, Copy)]
+struct SlotInfo {
+    value: Value,
+    offs: i32,
+    size: u32,
+}
+
+// Find out if the range `[offs, +size)` overlaps the with range in `si`.
+fn any_overlap(si: &SlotInfo, offs: i32, size: u32) -> bool {
+    let a_offs = si.offs as i64;
+    let a_size = si.size as i64;
+    let b_offs = offs as i64;
+    let b_size = size as i64;
+    let no_overlap = a_offs + a_size <= b_offs || b_offs + b_size <= a_offs;
+    !no_overlap
+}
+
+// Find, in `banks`, the register bank that `reg` lives in, and return the
+// lower limit and size of the bank.  This is so the caller can conveniently
+// iterate over all `RegUnits` in the bank that `reg` lives in.
+fn find_bank_limits(banks: &[RegBank], reg: RegUnit) -> (RegUnit, u16) {
+    for rb in banks.iter() {
+        if reg >= rb.first_unit && reg < rb.first_unit + rb.units {
+            return (rb.first_unit, rb.units);
+        }
+    }
+    // We should never get here, since `reg` must come from *some* RegBank.
+    panic!("find_regclass_limits: reg not found");
+}
+
+// The redundant-fill remover processes each EBB independently.  It makes a
+// single forward pass over each EBB, carrying along a running environment,
+// `AvailEnv`, that maps `RegUnits` to an `Option<SlotInfo>`.
+//
+// If the a triplet `(regunit, offset, size)` appears in the environment, this
+// indicates that, at this point, `regunit` and the stack slot denoted by
+// `(offset, size)` have the same value.  We also keep track of the SSA
+// `Value` associated with the `reg`, because that is necessary to create
+// `copy` instructions that read `reg`.
+//
+// Invariants for the environment:
+//
+// The environment may have multiple different registers bound to the same
+// stack slot (that is, offset+size pair).  That's OK, and reflects the
+// reality that those two registers contain the same value.
+//
+// The `value` fields must however be unique: it isn't allowable to have two
+// entries in the environment that have the same `value` field.  To do so
+// would be to imply that that SSA value has two different locations, since
+// each environment entry is associated with a different register, and so this
+// would imply that the Value resided in both of those registers.  Function
+// `invariants_ok` checks this.
+//
+// It is also the case that the set of `(offset, size)` segments in the
+// environment must be non-overlapping.  `invariants_ok` doesn't check that,
+// though.
+
+struct AvailEnv {
+    map: Vec<Option<SlotInfo>>,
+    size: usize,
+}
+
+impl AvailEnv {
+    // Create a new one.
+    fn new(size: usize) -> Self {
+        let mut env = AvailEnv {
+            map: Vec::<Option<SlotInfo>>::new(),
```

Comment:
Can you please store this in the `PeepHole` struct instead and clear it when calling `PeepHole::clear` to reduce the amount of allocations. The purpose of the `Context` is to cache heap allocated memory for use between compilation of two functions.

Reply:
Done.  As a side comment, we should really profile CL with DHAT to get a system-wide picture of the effectiveness, and turnover rate, of its heap usage.